src/model/validation/location.ts

```diff
@@ -0,0 +1,170 @@
+import { ASTNode, Kind, Location } from 'graphql';
+import { ProjectSource } from '../../project/source';
+import { getLineAndColumnFromPosition } from '../../schema/schema-utils';
+
+export class SourcePosition {
+    /**
+     * @param {number} offset the zero-based character offset within the source
+     * @param {number} line the one-based line number
+     * @param {number} column the one-based column number
+     */
+    constructor(
+        public readonly offset: number,
+        public readonly line: number,
+        public readonly column: number,
+    ) {}
+}
+
+export type LocationLike = MessageLocation | Location | ASTNode;
+
+export class MessageLocation {
+    public readonly sourceName: string;
+
+    // ! - we set this as a property
+    public readonly source!: ProjectSource;
+
+    constructor(
+        source: string | ProjectSource,
+        readonly _start: SourcePosition | number,
+        readonly _end: SourcePosition | number,
+    ) {
+        if (source instanceof ProjectSource) {
+            Object.defineProperty(this, 'source', {
+                enumerable: false,
+                value: source,
+            });
```

Comment:
Why is this needed?

Reply:
I [didn't leave an explanation when I did it](https://github.com/AEB-labs/cruddl/commit/96436ca80845bde06e71cf0f5f3148a772bce622). I don't remember either.

Reply:
Okay you can leave it like that of course, I was just curious.